exercise-book/src/iterators.md

```diff
@@ -0,0 +1,324 @@
+use std::fs::File;
+use std::error::Error;
+
+fn main() -> Result<(), Box<dyn Error>> {
+    use crate::*;
+    let f = File::open("../exercise-templates/iterators/numbers.txt")?;
+    let reader = BufReader::new(f);
+
+    let numeric_lines = reader.lines()
+      .map(|l| l.unwrap())
+      .map(|s| s.parse::<i32>())
+      .filter(|s| s.is_ok())
+      .map(|l| l.unwrap().to_string())
+      .collect::<Vec<String>>();
+    println!("{:?}", numeric_lines);
+
+    Ok(())
+}
+```
+
+</details>
+
+### Step 4: Keep the odd numbers
+
+Use a [.filter()](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.filter) with an appropriate closure.
+
+<details>
+  <summary>Solution</summary>
+
+```rust
+#![allow(unused_imports)]
+use std::io::{BufRead, BufReader};
+use std::fs::File;
+use std::error::Error;
+
+fn main() -> Result<(), Box<dyn Error>> {
+    use crate::*;
+    let f = File::open("../exercise-templates/iterators/numbers.txt")?;
+    let reader = BufReader::new(f);
+
+    let odd_numbers = reader.lines()
+      .map(|l| l.unwrap())
+      .map(|s| s.parse())
+      .filter(|s| s.is_ok())
+      .map(|l| l.unwrap())
+      .filter(|num| num % 2 != 0)
+      .collect::<Vec<i32>>();
+
+    println!("{:?}", odd_numbers);
+
+    Ok(())
+}
+```
+
+</details>
+
+### Step 5: Add the odd numbers
+
+Take the odd numbers, `.collect()` into a vector, and add them using a [.fold()](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.fold).
+
+You will probably reach for a `.sum::<i32>()`, but `.fold()`s are common enough in idiomatic Rust that we wanted to showcase one here.
+
+<details>
+  <summary>Solution</summary>
+
+```rust
+#![allow(unused_imports)]
+use std::io::{BufRead, BufReader};
+use std::fs::File;
+use std::error::Error;
+
+fn main() -> Result<(), Box<dyn Error>> {
+    use crate::*;
+    let f = File::open("../exercise-templates/iterators/numbers.txt")?;
+    let reader = BufReader::new(f);
+
+    let result = reader.lines()
+      .map(|l| l.unwrap())
+      .map(|s| s.parse())
+      .filter(|s| s.is_ok())
+      .map(|l| l.unwrap())
+      .filter(|num| num % 2 != 0)
+      .collect::<Vec<i32>>()
+      .iter()
+      .fold(0, |acc, elem| acc + elem);
+    // Also works
+    //.sum::<i32>();
+
+    println!("{:?}", result);
+
+    Ok(())
+}
+```
+
+</details>
+
+### Step 6: Idiomatic Rust
+
+That first solution can be a *slog*. 
+
+Try writing a shorter solution using a [.filter_map()](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.filter_map).
+
+<details>
+  <summary>Solution</summary>
+
+```rust
+#![allow(unused_imports)]
+use std::io::{BufRead, BufReader};
+use std::fs::File;
+use std::error::Error;
+
+fn main() -> Result<(), Box<dyn Error>> {
+    use crate::*;
+    let f = File::open("../exercise-templates/iterators/numbers.txt")?;
+    let reader = BufReader::new(f);
+
+    let result = reader.lines()
+      .map(|l| l.unwrap())
+      .filter_map(|s| s.parse().ok())
```

Comment:
While I appreciate a gradual progression on the solution, this particular sequence of steps (Step 2 to Step 6) goes against of what we teach people on trainings. They don't "`unwarp() now, refactor later`". Instead as soon as trainee sees a `Result`, they will think how to properly handle it. The final solution won't have any `unwrap / expect` calls in code. So I would give a different hint early: about the fact that in this problem we skip over "bad" lines, and then use this fact to introduce `filter_map` as an important method.